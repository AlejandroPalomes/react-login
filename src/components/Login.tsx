import { Redirect, Link } from "react-router-dom";
// import React, { useState, useEffect, useContext } from 'react';  //? to use with react context
import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
// import '../App.css';
import Button from './Button';
import LoginInput from './LoginInput';
import Logo from './Logo';
import Card from './Card';
import ErrorMessage from './ErrorMessage';

// import authContext from '../auth-context'; //? to use with react context

const LogoContainer = styled.div`
    margin-bottom: 50px;
    padding-top: 0;
    display: flex;
    align-items: center;
    justify-content: center;
`;

const StyledLogo = styled(Logo)`
    display: none;

    @media(min-width: 375px){
        display: inline;
    }
`;

const Name = styled.h1`
    font-size: 60px;
    font-family: "Permanent";
    color: rgb(109, 104, 255);
    margin: 0 0 0 10px;
`;

const StyledSpan = styled.span`
    font-size: 18px;
    color: rgb(22, 30, 46);
`;

const StyledLink = styled(Link)`
    color: rgb(109, 104, 255);
    font-weight: bold;
    text-decoration: none;
`;

interface props {
    setToken: Function
}

const Login = ({setToken}:props) => {

    const [disable, setDisable] = useState(true);
    const [email, setEmail] = useState('');
    const [emailErrorMessage, setEmailErrorMessage] = useState('');
    const [password, setPassword] = useState('');
    const [passwordErrorMessage, setPasswordErrorMessage] = useState('');
    const [redirect, setRedirect] = useState(false);
    const [waitingResponse, setWaitingResponse] = useState(false);
    // const {token, setToken, clearLocalStorage} = useContext(authContext); //? to use with react context

    interface ServerResponse {
        newToken: string
    }

    const checkLoginButton = () => setDisable((email.length && password.length) ? false : true);
    const updateEmail = (e: string) => setEmail(e);
    const updatePassword = (e: string) => setPassword(e);
    const checkInputs = () => {
        if (checkEmail() && checkPassword()) requestUser()
    };

    const checkEmail = () => {
        const emailRegEx = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

        if (emailRegEx.test(email)) {
            setEmailErrorMessage('');
            return true;
        } else {
            setEmailErrorMessage('Must be an email: email@domain')
            return false;
        }
    }

    const checkPassword = () => {
        const passwordRegEx = /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[^A-Za-z0-9]).{8,}$/;

        if (passwordRegEx.test(password)) {
            setPasswordErrorMessage('');
            return true;
        } else {
            setPasswordErrorMessage('Conditions: 8 char long, 1 lowercase, 1 uppercase, 1 number & 1 special character.')
            return false;
        }
    }

    const requestUser = () => {
        setWaitingResponse(true);
        setDisable(true);

        new Promise((resolve, reject) => {

                const mockResponse = {
                    newToken: 'ahdGFr59HfYn4j8S'
                }

                setTimeout(() => {
                    if (email === 'admin@email.com') password === 'Pa$$w0rd' ? resolve(mockResponse) : reject('pwd');
                    else reject('user');
                }, 1250);
            })
            .then((e: any) => handleResponse(e))
            .catch(e => handleServerError(e));
    }

    const handleResponse = ({ newToken }: ServerResponse) => {
        setToken(newToken);
        setWaitingResponse(false);
        setDisable(false);
        setRedirect(true);
    }

    const handleServerError = (message: string) => {
        if (message === 'user') setEmailErrorMessage('User not found.');
        else if (message === 'pwd') setPasswordErrorMessage('Sorry, wrong password.');
        else console.log(message);

        setWaitingResponse(false);
        setDisable(false);
    }

    useEffect(() => {
        checkLoginButton();
        // let isMounted = true;
        // if(isMounted) checkLoginButton();
        // return () =>{ isMounted = false};    //* This function is executed when component unmounts
    }, [email, password])

    return (
        <div className = "App">
            <header className = "App-header">
                { redirect && <Redirect to="/" /> }
                <Card
                    width = { 410 }
                    padding = { 50 }
                >
                    <LogoContainer>
                        <StyledLogo
                            size = { 75 }
                        />
                        <Name>ukiyo</Name>
                    </LogoContainer>
                    <LoginInput
                        hide = { false }
                        placeholder = 'Email'
                        parentFunction = { updateEmail }
                    />
                    <ErrorMessage
                        message = { emailErrorMessage }
                    />
                    <LoginInput
                        hide = { true }
                        placeholder = 'Password'
                        parentFunction = { updatePassword }
                    />
                    <ErrorMessage
                    message = { passwordErrorMessage }
                    />
                    <Button
                        disabled = { disable }
                        content = 'Login'
                        buttonAction = { checkInputs }
                        isLoading = { waitingResponse }
                    />
                </Card>
                <Card
                    width = { 410 }
                    padding = { 20 }
                >
                    <StyledSpan>New user? <StyledLink to = "/signup">Register here</StyledLink></StyledSpan>
                </Card>
            </header>
        </div>
    );
}

export default Login;