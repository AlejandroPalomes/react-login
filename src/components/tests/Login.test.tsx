
import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router, Switch, Route, Redirect } from "react-router-dom";
import '@testing-library/jest-dom';
import { render, fireEvent, waitFor } from '@testing-library/react';

import Login from '../Login';

describe('Login Component test suite', () => {


    const getter = 'property';
    const setter = jest.fn();
    const defaultProps = [getter, setter];
    const component = (
        <Router>
            <Login setToken={defaultProps}/>
        </Router>
    )
    const setup = () => render(component);

    it('Renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(component, div);
        ReactDOM.unmountComponentAtNode(div)
    })

    it('Should match snapshot', () => {
        const { container } = setup();
        expect(container).toMatchSnapshot();
    })

    it('Should return error message on email when it is not an email', async ()=>{
        const { getByText, getByPlaceholderText } = setup();
        const loginButton = getByText('Login');
        const emailInput = getByPlaceholderText('Email');
        const passwordInput = getByPlaceholderText('Password');
        fireEvent.change(emailInput, { target: { value: '23' } })
        fireEvent.change(passwordInput, { target: { value: '23' } })
        fireEvent.click(loginButton)

        expect(getByText('Must be an email: email@domain')).toBeDefined();
    })

    it('Should return error message on password when it does not match all conditions', async ()=>{
        const { getByText, getByPlaceholderText } = setup();
        const loginButton = getByText('Login');
        const emailInput = getByPlaceholderText('Email');
        const passwordInput = getByPlaceholderText('Password');
        fireEvent.change(emailInput, { target: { value: 'admin@email.com' } })
        fireEvent.change(passwordInput, { target: { value: '23' } })
        fireEvent.click(loginButton)

        expect(getByText('Conditions: 8 char long, 1 lowercase, 1 uppercase, 1 number & 1 special character.')).toBeDefined();
    })

    it('Should return error message on email when user not found', async ()=>{

        const { getByText, getByPlaceholderText } = setup();
        const loginButton = getByText('Login');
        const emailInput = getByPlaceholderText('Email');
        const passwordInput = getByPlaceholderText('Password');

        fireEvent.change(emailInput, { target: { value: 'user@email.com' } })
        fireEvent.change(passwordInput, { target: { value: 'Pa$$w0rd' } })
        fireEvent.click(loginButton)

        await waitFor(()=>{
            expect(getByText('User not found.')).toBeDefined();
        }, {timeout: 2000})
    })

    it('Should return error message on password when user found but password is not correct', async ()=>{

        const { getByText, getByPlaceholderText } = setup();
        const loginButton = getByText('Login');
        const emailInput = getByPlaceholderText('Email');
        const passwordInput = getByPlaceholderText('Password');

        fireEvent.change(emailInput, { target: { value: 'admin@email.com' } })
        fireEvent.change(passwordInput, { target: { value: 'Pa$$w0rdddd' } })
        fireEvent.click(loginButton)

        await waitFor(()=>{
            expect(getByText('Sorry, wrong password.')).toBeDefined();
        }, {timeout: 2000})
    })

    it('Should give access if correct email and password', async ()=>{

        const { getByText, getByPlaceholderText, queryByText } = setup();
        const loginButton = getByText('Login');
        const emailInput = getByPlaceholderText('Email');
        const passwordInput = getByPlaceholderText('Password');

        fireEvent.change(emailInput, { target: { value: 'admin@email.com' } })
        fireEvent.change(passwordInput, { target: { value: 'Pa$$w0rd' } })
        fireEvent.click(loginButton)

        await waitFor(()=>{
            expect(queryByText('Sorry, wrong password.')).toBeNull();
            expect(queryByText('User not found.')).toBeNull();
        }, {timeout: 2000})
    })
});