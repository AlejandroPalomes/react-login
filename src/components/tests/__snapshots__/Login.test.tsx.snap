// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component test suite Should match snapshot 1`] = `
<div>
  <div
    class="sc-fubCfw bPCeor"
  >
    <div
      class="sc-pFZIQ eKkxAB"
    >
      <h3>
        all your needs in one place
      </h3>
      <img
        alt="City connected"
        src="cityImage.svg"
        width="100%"
      />
    </div>
    <div
      class="sc-jrAGrp cKcZr"
    >
      <div
        class="sc-gsTCUz dDSAcv"
        width="410"
      >
        <div
          class="sc-kEjbxe efYBx"
        >
          <svg
            class="sc-iqHYGH jyGpMs"
            height="75"
            version="1.1"
            viewBox="0 0 512 512"
            width="75"
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
            y="0px"
          >
            <path
              class="sc-gKsewC fxHmnu"
              d="M256,512L256,512C114.6,512,0,397.4,0,256v0C0,114.6,114.6,0,256,0h0c141.4,0,256,114.6,256,256v0 C512,397.4,397.4,512,256,512z"
            />
            <path
              class="sc-iBPRYJ jAOHRc"
              d="M398.2,369.7c-6.6,0-12.6,2.4-17.5,6.2l-54-30.5c0.6-4.1,0.5-4.4,0-7.9l54-30.8c4.8,3.8,10.8,6.1,17.4,6.1 c15.7,0,28.4-12.8,28.4-28.4c0-15.7-12.8-28.4-28.4-28.4s-28.4,12.8-28.4,28.4c0,1.3,0.2,2.5,0.4,3.8l-54,30.8 c-4.8-3.8-10.8-6.1-17.4-6.1c-15.7,0-28.4,12.8-28.4,28.4c0,15.7,12.8,28.4,28.4,28.4c6.5,0,12.4-2.3,17.2-6l54.3,30.6 c-0.2,1.2-0.4,2.5-0.4,3.8c0,15.7,12.8,28.4,28.4,28.4c15.7,0,28.4-12.7,28.4-28.4S413.9,369.7,398.2,369.7z"
            />
            <path
              class="sc-iBPRYJ jAOHRc"
              d="M327.1,170.1v-56.3c0-7.8-6.4-14.2-14.2-14.2c-7.8,0-14.2,6.4-14.2,14.2v28.5l-54-52.8 c-5.5-5.4-14.4-5.4-19.9,0l-132.9,130c-4.1,4.1-6.4,9.5-6.4,15.2c0,11.8,9.6,21.3,21.3,21.3H128v78.2c0,19.6,15.9,35.6,35.6,35.6h64 c7.8,0,14.2-6.4,14.2-14.2s-6.4-14.2-14.2-14.2h-64c-3.9,0-7.1-3.2-7.1-7.1v-92.4c0-7.8-6.4-14.2-14.2-14.2h-18.1l110.5-108.1 l123,120.3c5.6,5.5,14.7,5.4,20.1-0.2c5.5-5.6,5.4-14.6-0.2-20.1L327.1,170.1z"
            />
          </svg>
          <h1
            class="sc-crrsfI jVKYIx"
          >
            ukiyo
          </h1>
        </div>
        <div
          class="sc-hKgILt dRyOZN"
        >
          <input
            class="sc-eCssSg jqOfMV"
            placeholder="Email"
            type="text"
          />
          
        </div>
        <div
          class="sc-hKgILt dRyOZN"
        >
          <input
            class="sc-eCssSg jqOfMV"
            placeholder="Password"
            type="password"
          />
          <button
            class="sc-jSgupP kmEZLF"
          >
            Show
          </button>
        </div>
        <button
          class="sc-bdfBwQ bbjigB"
          disabled=""
        >
          <span>
            Login
          </span>
        </button>
        <a
          class="sc-kstrdz iJyhPM"
          href="/forgotten"
        >
          Forgot password?
        </a>
      </div>
      <div
        class="sc-gsTCUz fWmxGQ"
        width="410"
      >
        <span
          class="sc-dQppl vnUDC"
        >
          New user? 
          <a
            class="sc-bqyKva gaJqUg"
            href="/signup"
          >
            Register here
          </a>
        </span>
      </div>
    </div>
  </div>
</div>
`;
